numpy/core/src/umath/scalarmath.c.src

```diff
@@ -1353,36 +1619,52 @@ static NPY_INLINE PyObject *
  *         long, ulong, longlong, ulonglong,
  *         half, float, double, longdouble,
  *         cfloat, cdouble, clongdouble#
+ * #Name = Byte, UByte, Short, UShort, Int, UInt,
+ *         Long, ULong, LongLong, ULongLong,
+ *         Half, Float, Double, LongDouble,
+ *         CFloat, CDouble, CLongDouble#
  * #simp = def*10, def_half, def*3, cmplx*3#
  */
+#define IS_@name@
+
 static PyObject*
 @name@_richcompare(PyObject *self, PyObject *other, int cmp_op)
 {
     npy_@name@ arg1, arg2;
-    int out=0;
-
-    RICHCMP_GIVE_UP_IF_NEEDED(self, other);
+    int out = 0;
 
-    switch(_@name@_convert2_to_ctypes(self, &arg1, other, &arg2)) {
-    case 0:
-        break;
-    case -1:
-        /* can't cast both safely use different add function */
-    case -2:
-        /* use ufunc */
-        if (PyErr_Occurred()) {
-            return NULL;
-        }
-        return PyGenericArrType_Type.tp_richcompare(self, other, cmp_op);
-    case -3:
-        /*
-         * special case for longdouble and clongdouble
-         * because they have a recursive getitem in their dtype
-         */
-        Py_INCREF(Py_NotImplemented);
-        return Py_NotImplemented;
+    /*
+     * Extract the other value (if it is compatible).
+     */
+    int res = convert_to_@name@(other, &arg2);
+    switch (res) {
+        case CONVERSION_ERROR:
+            return NULL;  /* an error occurred (should never happen) */
+        case DEFER_TO_OTHER_KNOWN_SCALAR:
+            Py_RETURN_NOTIMPLEMENTED;
+        case CONVERSION_SUCCESS:
+            break;  /* successfully extracted value we can proceed */
+        case OTHER_IS_UNKNOWN_OBJECT:
+#if defined(IS_longdouble) || defined(IS_clongdouble)
+            Py_RETURN_NOTIMPLEMENTED;
+#endif
+            RICHCMP_GIVE_UP_IF_NEEDED(self, other);
+        case PROMOTION_REQUIRED:
+            return PyGenericArrType_Type.tp_richcompare(self, other, cmp_op);
+        case OTHER_IS_SUBCLASS:
+            /*
+             * Success, but defer if not a forward operator.  This implements
+             * the asymmetry of `__add__` and `__radd__` for subclassing.
+             * (A subclass replacing `__add__`/`__radd__` has precedence. It
+             * also means that such a subclass has to always take charge.)
+             */
+            if ((void *)Py_TYPE(other)->tp_richcompare != (void *)@name@_richcompare) {
+                Py_RETURN_NOTIMPLEMENTED;
+            }
```

Comment:
In both these cases, can you let the `self.tp_richcompare` handle it? That way if a subclass has overridden `tp_richcompare` it should JustWork (or return NotImplemented).

Reply:
The case that this "fixes" is when you have two subclasses:
```
class myf32(np.float32):
    pass

class my_customf32(np.float32):
    def __eq__(self, other):
        return "something strange"
```
The above logic allows that:
```
myf32(3) == my_customf32(3)
```
defers to `my_customf32`.  The reason is that `myf32` will see that `my_customf32` has custom handling and ask it to do the operation.

If we continue, we return the NumPy behaviour there always (meaning that the operation result will depend on order).  If we return `NotImplemented` always, we would return `False` for the case of:
```python
class myf32_2(np.float32):
    pass

myf32(1) == myf32_2(1)
```
Where two _different_ subclasses which both inherit the comparison are compared.

Reply:
For most binary operators the rules would be clear, we would defer in `__add__`, but not in `__radd__`.  But comparisons don't provide the op vs. rop information properly :/

Reply:
> If we continue, we return the NumPy behaviour there always (meaning that the operation result will depend on order). 

This is consistent with Python semantics.

Reply:
>>> class A(int):
...     def __eq__(self, other):
...         return 'A'
... 
>>> class B(int):
...    pass
... 
>>> B(3) == A(3)
True
>>> A(3) == B(3)
'A'

Reply:
Arrg, you are right... seems even Python itself doesn't even try to fix this for the other binary operators, the same is true for `__add__` for these builtin classes.

In that case, we can punt and just hope we are fine.  This branch may be more correct, but there is no point in trying harder than Python `int` itself.

Reply:
This is probably documented somewhere. It is different when the comparison is between instances of a class and a subclass:
```
>>> class A():
...     pass
... 
>>> class B(A):
...    def __eq__(self, other):
...       return 'B'
... 
>>> A() == B()
'B'
>>> B() == A()
'B'
```

Reply:
Yeah, for direct subclasses Python will ensure the correct order by just calling the subclass first.

There is also this funny assymetry, that if you write Python code at least, you can make `__radd__` mora aggressive than `__add__` to allow a subclass that coerces other subclasses more aggressively to take charge.
But, making that assymetry work in C is very annoying, and it is pretty esoteric anyway.  If even Python `int` bails on it, there is no reason we should not...

Reply:
OK, removed that additional logic to try to make seperate subclasses work.  I think the remaining "is forward" logic for binops is very similar to what we already have for arrays, and should be fine for PyPy (I cannot think of a failure path, and I don't think it is a change to begin with now).

Reply:
:100: